✨ [feat] #32 투두 생성 API 구현

yeeun0702 opened the pull request:
## 🍞 Issue

Closes #32 

사용자가 카테고리 내에서 할 일을 생성할 수 있는 기능을 구현했습니다.
할 일은 날짜, 내용, (선택적으로) 시작 시간을 포함하며, 특정 카테고리에 속하도록 되어 있습니다.


## 🥐 Todo
- [ ] TodoCreateReq 요청 DTO 및 TodoCreateRes 응답 DTO 정의
- [ ] 도메인 모델(Todo) 및 엔티티(TodoEntity) 정의 및 생성 팩토리 메서드 추가
- [ ] TodoService, TodoFacade, TodoSaver 계층을 통해 생성 로직 구현
- [ ] POST /todos API 컨트롤러 구현


## 🧇 Details
 `Todo` 생성 시 **JPA 연관관계 설정**을 위해 `CategoryEntity`를 직접 조회해 사용했습니다.
이는 `TodoEntity`에 `@ManyToOne`으로 매핑된 필드에 category를 명시적으로 주입해주기 위함입니다.

```java
// TodoEntity 생성 시
TodoEntity.forCreate(todo, categoryEntity);
```

### 🤔 이 방식이 맘에 걸리는 이유:

* Facade 계층에서 굳이 Entity를 직접 넘기는 게 도메인 의도를 흐릴 수 있을 것 같음.
* 원래는 도메인 모델(`Category`)만 사용하고 싶었지만, 연관 매핑 특성상 어쩔 수 없이 사용


## 🖼 Postman Screenshots
<img width="1039" height="613" alt="image" src="https://github.com/user-attachments/assets/61e18ee5-071d-41f7-984f-4b2f5516106d" />


## 🍩 Reviewer Notes
기능명세서 참고해서 예외처리 조금 더 꼼꼼하게 하겠습니당 !
카테고리 최하단 순서지정, 날짜 형식 안 맞는 경우, 글자 수 관련 처리, 시간 디폴트 설정(서버 해당사항X) 우선 이 정도 예외처리 했는데 혹시 놓친 게 있다면 알려주세요 !! 그리고 지금 방식이 좀 이상한 거 같은데 (멀티모듈 어렵네용 ㅠㅠ) 먼가 더 좋은 방식이 있으까요 ?? 추천받고 리팩하는게 빠를 듯 합니다 ... ㅎㅎ

yeeun0702 committed:
✨ [feat] #32 Todo 생성 요청/응답 DTO 및 API 구현

yeeun0702 committed:
✨ [feat] #32 TodoSaver 및 Facade 계층 구현

yeeun0702 committed:
✨ [feat] #32 Todo 도메인 및 서비스 구현

yeeun0702 committed:
✨ [feat] #32 CategoryRetriever에 Entity 직접 조회 메서드 추가

- Todo 생성 시, 카테고리 소유자 검증 및 연관 매핑을 위해 CategoryEntity 필요
- findEntityByIdAndUserId() 메서드 별도 추가

yeeun0702 committed:
✨ [feat] #32 날짜 형식 오류에 대한 예외 처리 추가

choyeongju commented on bbangzip-api/src/main/java/org/sopt/category/dto/req/CategoryReorderReq.java:
```diff
@@ -0,0 +1,12 @@
+
+public record CategoryReorderReq(
+        @NotEmpty(message = "카테고리 순서 리스트는 비어 있을 수 없습니다.")
+        @PositiveIdList
```
오 이게 그 커스텀 어노테이션이군요!!

choyeongju commented on bbangzip-api/src/main/java/org/sopt/todo/controller/TodoController.java:
```diff
@@ -0,0 +1,36 @@
+    ) {
+        Long dummyUserId = 1L;
+
+        return ResponseEntity
+                .status(HttpStatus.CREATED)
+                .body(BaseResponse.success(SuccessCode.CREATED, todoService.createTodo(dummyUserId, todoCreateReq)));
+    }
```
혹시 요거 .body() 내부에 BaseResponse 이쪽 안넣으면 동작 안할까용?!

yeeun0702 replied:
### 🔍 예상되는 문제 포인트
https://github.com/Team-BBANGZIP/BBANGZIP-SERVER/blob/02111d0dc85ef4818e54e9728fe4ea01826e42a1/bbangzip-api/src/main/java/org/sopt/common/ResponseWrappingAdvice.java#L34-L36

ResponseEntity는 자동 래핑 대상에서 제외되기 때문에, body instanceof ResponseEntity 조건에 의해 ResponseEntity를 반환하면 BaseResponse.success(...)로 감싸지지 않고 그대로 응답됩니다.

그 결과, HTTP status는 정상적으로 적용되지만, 응답 본문의 code 값은 기대했던 20100이 아니라 기본값인 20000으로 내려가게 되더라고요. 즉, 우리가 의도한 code, message 등 통일된 응답 포맷이 적용되지 않는 거죠!

그래서 `ResponseWrappingAdvice` 를 수정하는 것도 생각했으나, 공수가 꽤 클 거 같아서 지금처럼 해두었어요 !!!

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/category/exception/CategoryCoreErrorCode.java:
```diff
@@ -11,7 +11,11 @@ public enum CategoryCoreErrorCode implements ErrorCode {
+  INVALID_CATEGORY_ORDER(HttpStatus.BAD_REQUEST, 40003, "카테고리 순서 정보가 유효하지 않습니다."),
+  INVALID_CATEGORY_ORDER_MISMATCH(HttpStatus.BAD_REQUEST, 40004, "카테고리 수와 요청된 ID 수가 일치하지 않습니다."),
+  INVALID_CATEGORY_ORDER_DUPLICATE(HttpStatus.BAD_REQUEST, 40005, "중복된 ID가 존재합니다."),
+  INVALID_CATEGORY_ORDER_MISSING(HttpStatus.BAD_REQUEST, 40006, "누락된 ID가 존재합니다.");
```
꼼꼼한 케이스별 예외처리 최고네요>,,<

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/todo/domain/TodoEntity.java:
```diff
@@ -56,4 +57,29 @@ public class TodoEntity extends BaseTimeEntity {
+        return entity;
+    }
+
+
```
요기 띄어쓰기 하나 삭제 부탁해요!

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/todo/facade/TodoSaver.java:
```diff
@@ -0,0 +1,30 @@
+@RequiredArgsConstructor
+public class TodoSaver {
+
+    private final TodoRepository todoRepository;
+    private final CategoryFacade categoryFacade;
```
Saver는 영속화만 담당하고, 소유권 검증은 Service/Façade 계층에서 끝내고 들어오면 좋을 것 같다는 생각이 들어요!

yeeun0702 replied:
https://github.com/Team-BBANGZIP/BBANGZIP-SERVER/pull/33/commits/8d5515fa9168761f882015641c756e8cee7a7c7b 

Service/Façade 계층에서 검증을 모두 끝내고, Saver는 영속화만 담당하도록 했어용 ,,, Command 패턴 도입도 고민해봤지만, 일단은 파라미터 방식으로 구현했습니다. 이런 방향성이 맞을까요 ?! DDD 관점에서 더 개선할 만한 부분이 있다면 조언 부탁드립니다 😊

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/todo/facade/TodoSaver.java:
```diff
@@ -0,0 +1,30 @@
+
+    public Todo save(final Todo todo) {
+
+        Long userId = todo.getCategory().getUserId();
+        Long categoryId = todo.getCategory().getId();
```
도메인 객체에서 userId, categoryId를 꺼내는 로직은 상위 계층에서 검증까지 같이 처리하고, Saver는 id만 받게 단순화하면 깔끔할 것 같습니당

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/todo/facade/TodoSaver.java:
```diff
@@ -0,0 +1,30 @@
+        Long categoryId = todo.getCategory().getId();
+
+        CategoryEntity categoryEntity = categoryFacade.getEntityByIdAndUserId(categoryId, userId);
+        TodoEntity entity = TodoEntity.forCreate(todo, categoryEntity);
```
TodoEntity 생성 시 CategoryEntity를 직접 넘기지 않고, categoryId만 넘기는 방법이 좋을 것 같아요!

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/todo/facade/TodoSaver.java:
```diff
@@ -0,0 +1,30 @@
+        TodoEntity entity = TodoEntity.forCreate(todo, categoryEntity);
+        TodoEntity saved = todoRepository.save(entity);
+
+        return saved.toDomain();
```
Saver는 보통 Entity 저장까지만 하고 변환은 상위에서 담당해도 좋을 것 같아요~!!

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/todo/facade/TodoSaver.java:
```diff
@@ -0,0 +1,30 @@
+        Long userId = todo.getCategory().getUserId();
+        Long categoryId = todo.getCategory().getId();
+
+        CategoryEntity categoryEntity = categoryFacade.getEntityByIdAndUserId(categoryId, userId);
```
프록시로 조회해도 괜찮을거같아요!!

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/category/facade/CategoryUpdater.java:
```diff
@@ -0,0 +1,122 @@
+
+        for (CategoryEntity entity : categoryEntities) {
+            Integer newOrder = idToOrder.get(entity.getId());
+            if (newOrder != null && !newOrder.equals(entity.getOrder())) { // 순서가 변경된 경우에만 업데이트
+                entity.update(
+                        entity.getName(),
+                        entity.getColor(),
+                        entity.isStopped(),
+                        newOrder // 새로운 정렬 순서
+                );
```
name/color/isStopped는 변경 안 되는데 매번 넘기는 게 조금 부자연스럽다고 느껴져요..! 0>0

choyeongju commented on bbangzip-domain/src/main/java/org/sopt/category/facade/CategoryUpdater.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Transactional
+    public Category updateCategory(Long categoryId, Long userId, String newName, CategoryColor newColor, boolean newIsStopped) {
+        CategoryEntity categoryEntity = categoryRepository.findByIdAndUserId(categoryId, userId)
+                .orElseThrow(() -> new CategoryNotFoundException(CATEGORY_NOT_FOUND));
```
Updater에서 findByIdAndUserId 같은 조회/검증 로직까지 들고 있네요. 보통 이런 검증 책임은 Retriever 쪽에 두고, Updater는 순수하게 상태 변경만 담당하는 게 더 자연스럽게 느껴질 것 같아요!

전반적으로 Facade 내부 클래스들끼리 역할 경계가 살짝 섞여 있을 가능성도 있을 것 같아서, 다른 Facade의 조회 책임을 끌어다 쓰지 않았는지 점검해보면 좋을 것 같아요 🙏!!

yeeun0702 replied:
맞아요 ㅠㅠ 사실 Updater에서 findByIdAndUserId까지 하고 있는 게 애초에 책임 분리를 의식적으로 하려고 했던 구조랑 다르게 돼버렸네요... 🫠 

원래는 말씀해주신 것처럼 Updater는 상태 변경만 담당하고, 조회/검증은 Retriever나 Validator에서 처리하고 Facade(Service)에서 조합하는 식으로 가야 했는데, 기능 구현에 집중하면 무의식적으로 서비스 레이어 역할까지 Updater 안에서 처리하게 되는 것 같아요. 투두는 의식적으로 책임분리를 하는걸루 노력해보고 ,,, 우선은 위에 부자연스럽다고 하신 내용만 https://github.com/Team-BBANGZIP/BBANGZIP-SERVER/pull/33/commits/bf02ae7baf82a7261f41daafa7f5e4b43284da9a 반영했구 나머진 나중에 분리해볼게요 !!!

choyeongju reviewed (commented):
우와 고봉밥이에요.. 보는데 오래걸렸다
구현하느라 수고하셧습니다 ㅠㅠ!!!

yeeun0702 committed:
♻️ [refac] #32 카테고리 순서 변경만 처리하도록 updateOrder 구현

yeeun0702 committed:
♻️ [refac] #32 TodoSaver 책임 단순화 및 영속화 구조 개선

- Saver는 영속화만 담당, 검증 및 변환은 Service/Façade 계층에서 처리
- Category 조회 시 프록시(getReferenceById) 사용
